### ISSUE TITLE
Where there are multiple projects opend in one nvim instance, null-ls will only start one instance at the first project root, and other project's local configuration files cannot be loaded.

### ISSUE THREAD

**Issue by milanglacier** [jose-elias-alvarez/null-ls.nvim#1134]
### FAQ

- [X] I have checked the [FAQ](https://github.com/jose-elias-alvarez/null-ls.nvim#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues?q=is%3Aissue) and there are no issues with the same problem.

### Neovim Version

NVIM v0.8.0-dev-1092-g89b9eab63-dirty

### Operating System

macOS

### Minimal config

```Lua
It's not related to config files.
```

### Steps to reproduce

1. Open files in one project.
2. Open another file in another project.

We can simply assume the two projects are two git projects.
Suppose the first project is a lua project, and the second is a python project,
I have some project local configuration files (for example `.flake8`),
then open files in the second project will result in, say the following diagnostic errors:

<img width="630" alt="Screen Shot 2022-09-24 at 04 16 27" src="https://user-images.githubusercontent.com/45728125/192087973-1bd581f2-0cbc-4800-8454-6cf52bdbf78a.png">

Because `flake8` spawned by null-ls does not start at the correct project root.

### Expected behavior

Like what `lspconfig` did when there are multiple projects are detected, spawn multiple lsps under different root directory.

### Actual behavior

Null-ls will only spawn one lsp in the first root directory.
Or if there is only one project in this nvim instance, the null-ls works fine, as the project root of null-ls is correct.

### Debug log

Not related

### Help

Yes, but I don't know how to start. I would need guidance

### Requirements

- [X] I have read and followed the instructions above and understand that my issue will be closed if I did not provide the required information.

**Comment by milanglacier** [comment]
https://github.com/jose-elias-alvarez/null-ls.nvim/blob/8af89c5fa2b732aaa9c3bf8aed95bccc9c4ce295/lua/null-ls/config.lua#L98

Is it possible to maintain an additional lua table which records the working directory null-ls launches.
And in an BufEnter autocmd to detect if there is a new project (new root), if true, add a new entry into this table and launch a new null-ls instance.

**Comment by jose-elias-alvarez** [comment]
Yep, this is currently not supported - see #1120 for discussion on why this is and what can be done about it. This isn't a workflow I personally use, but I'm happy to review / provide support with PRs improving behavior here. I think the approach you suggest is the right one, though there's a few other areas that assume only a single null-ls client will ever exist that also have to be updated. 

Also, for anyone who stumbles on this issue before it's solved, here is a workaround:

```lua
local flake8 = require("null-ls").builtins.diagnostics.flake8.with({
    cwd = function()
        return vim.loop.cwd()
    end,
})
```

**Comment by milanglacier** [comment]
Hi! Thanks for your suggestion. I would definitely will to make the contribution.

But I am thinking about that: there are some sources which don't need to respect project local settings, for example: `proselint`, `codespell`, `shellcheck`, then if we spawn multiple null-ls instances, then those cli program will be also spawned multiple times. Do we need to take care of it? Just leave it as it is?

**Comment by jose-elias-alvarez** [comment]
> But I am thinking about that: there are some sources which don't need to respect project local settings, for example: proselint, codespell, shellcheck, then if we spawn multiple null-ls instances, then those cli program will be also spawned multiple times. Do we need to take care of it? Just leave it as it is?

I don't think the behavior will change for built-in sources without local configuration (and I think the maintenance burden of tracking which sources do and don't need multiple clients is too high).

**Comment by milanglacier** [comment]
> Yep, this is currently not supported - see #1120 for discussion on why this is and what can be done about it. This isn't a workflow I personally use, but I'm happy to review / provide support with PRs improving behavior here. I think the approach you suggest is the right one, though there's a few other areas that assume only a single null-ls client will ever exist that also have to be updated.
> 
> Also, for anyone who stumbles on this issue before it's solved, here is a workaround:
> 
> ```lua
> local flake8 = require("null-ls").builtins.diagnostics.flake8.with({
>     cwd = function()
>         return vim.loop.cwd()
>     end,
> })
> ```

I have tried this in my config and it does not work. It seems that the only way to fix it is to allow null-ls to spawn multiple instances under different project root.

**Comment by milanglacier** [comment]
A temporary workaround would be:

```lua
    local null_ls = require 'null-ls'
    local util = require 'null-ls.utils'
            null_ls.builtins.diagnostics.flake8.with {
                command = mypath.flake8,
                cwd = function()
                    return util.root_pattern('.git', 'Makefile', 'pyproject.toml', '.flake8')(vim.fn.expand '%:h')
                end,
            },

```

**Comment by jose-elias-alvarez** [comment]
I see, thanks for mentioning the workaround for the plugin. We should make sure that whatever solution we come up with also handles the interaction, since I imagine a good percentage of users who want this feature are using the plugin (or a similar one).

**Comment by jose-elias-alvarez** [comment]
FYI: while the original issue still exists and will only be fixed by creating multiple null-ls clients, #1172 should fix the specific issue for `flake8`.
